src/runtime_src/core/common/usage_metrics.cpp

```diff
@@ -0,0 +1,515 @@
+      bpt::ptree dev;
+      dev.put("device_index", std::to_string(dev_id));
+      dev.put("bdf", dev_metrics.bdf);
+      dev.put("bos_peak_count", std::to_string(dev_metrics.bo_peak_count));
+
+      // add global bos
+      dev.add_child("global_bos", get_bos_ptree(dev_metrics.global_bos_met));
+
+      // add hw ctx info
+      dev.add_child("hw_context", get_hw_ctx_ptree(dev_metrics.hw_ctx_vec));
+
+      dev_array.push_back(std::make_pair("device", dev));
+    }
+
+    thread_array.add_child("thread " + std::to_string(t_count), dev_array);
+    t_count++;
+  }
+
+  print_json(thread_array);
+}
+
+// Create specific logger if ini option is enabled
+static std::shared_ptr<base_logger>
+get_logger_object()
+{
+  if (xrt_core::config::get_usage_metrics_logging())
+    return std::make_shared<usage_metrics_logger>();
+
+  return std::make_shared<base_logger>();
+}
+
+// Per thread logger object  
+std::shared_ptr<base_logger>
+get_usage_metrics_logger()
+{
+  static thread_local auto usage_logger_object = get_logger_object();
+  return usage_logger_object;
+}
+
+usage_metrics_logger::
+usage_metrics_logger() : map_ptr(usage_metrics_map)
+{
+  thread_count++;
+}
+
+usage_metrics_logger::
+~usage_metrics_logger()
+{
+  {
+    std::lock_guard<std::mutex> lk(m);
+    thread_count--;
+
+    // push this threads usage metrics to global map
+    // in thread safe manner
+    (*map_ptr)[std::this_thread::get_id()] = std::move(m_dev_map);
+  }
+
+  if (thread_count == 0) {
+    // print usage metrics log after all threads are destroyed
+    print_usage_metrics();
+  }
+}
+
+static device_metrics*
+get_device_metrics(device_metrics_map& dev_map, device_id dev_id)
+{
+  auto map_it = dev_map.find(dev_id);
+  return map_it == dev_map.end()
+    ? nullptr
+    : &(*map_it).second;
+}
+
+void
+usage_metrics_logger::
+log_device_info(const xrt_core::device* dev)
+{
+  auto dev_id = dev->get_device_id();
+  if (!get_device_metrics(m_dev_map, dev_id)) {
+    // initialize map with this device index
+    m_dev_map[dev_id] = {};
+    try {
+      auto bdf = xrt_core::query::pcie_bdf::to_string(xrt_core::device_query<xrt_core::query::pcie_bdf>(dev));
+      m_dev_map[dev_id].bdf = bdf;
+    }   
+    catch (...) {}
+  }
+}
+
+void 
+usage_metrics_logger::
+log_hw_ctx_info(const xrt::hw_context_impl* hwctx_impl)
+{
+  try {
+    auto hw_ctx = 
+        xrt_core::hw_context_int::create_hw_context_from_implementation(const_cast<xrt::hw_context_impl*>(hwctx_impl));
+    
+    auto hwctx_handle = static_cast<xrt_core::hwctx_handle*>(hw_ctx);
+    auto dev_id = xrt_core::hw_context_int::get_core_device(hw_ctx)->get_device_id();
+    auto uuid = hw_ctx.get_xclbin_uuid();
+
+    // dont log if device didn't match
+    auto dev_metrics = get_device_metrics(m_dev_map, dev_id);
+    if (!dev_metrics)
+      return;
+
+    // log if this entry is not logged before
+    if (!get_metrics<hw_ctx_metrics, xrt_core::hwctx_handle*>(dev_metrics->hw_ctx_vec, hwctx_handle))
```

Comment:
Template types are automatically inferred when you are passing all arguments.
```suggestion
    if (!get_metrics(dev_metrics->hw_ctx_vec, hwctx_handle))
```